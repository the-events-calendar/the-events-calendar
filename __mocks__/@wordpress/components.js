/**
 * External dependencies
 */
import { noop } from 'lodash';

export const withAPIData = () => noop;
export const Spinner = () => "🏃‍♂️";
export const Modal = ( { title, children } ) => (
	<div>
		<span>{ title }</span>
		<span>{ children }</span>
	</div>
);
export const Dashicon = ( { className, icon } ) => <span className={ className }>{ icon }</span>;
export const Dropdown = () => <span>Dropdown</span>;
export const Tooltip = () => <span>Tooltip</span>;
export const PanelBody = ({ children }) => <span className="PanelBody">{ children }</span>
