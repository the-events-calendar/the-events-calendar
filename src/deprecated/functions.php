<?php
/**
 * register a meta group
 *
 * @uses Tribe__Events__Meta_Factory::register()
 * @deprecated 4.3
 *
 * @param string $meta_group_id
 * @param array  $args
 *
 * @return bool $success
 */
function tribe_register_meta_group( $meta_group_id, $args = [] ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	// setup default for registering a meta group
	$defaults = [ 'register_type' => 'meta_group', 'register_overwrite' => true ];

	// parse the $default and $args into the second param for registering a meta item
	return Tribe__Events__Meta_Factory::register( $meta_group_id, wp_parse_args( $args, $defaults ) );
}

/**
 * register a meta item
 *
 * @uses Tribe__Events__Meta_Factory::register()
 * @deprecated 4.3
 *
 * @param int   $meta_id
 * @param array $args
 *
 * @return bool $success
 */
function tribe_register_meta( $meta_id, $args = [] ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	return Tribe__Events__Meta_Factory::register( $meta_id, $args );
}

/**
 * Get the meta group.
 *
 * @deprecated 4.3
 *
 * @param      $meta_group_id
 * @param bool $is_the_meta
 *
 * @return bool|mixed|void
 */
function tribe_get_meta_group( $meta_group_id, $is_the_meta = false ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	do_action( 'tribe_get_meta_group', $meta_group_id, $is_the_meta );

	$type = 'meta_group';

	// die silently if the requested meta group is not registered
	if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_group_id, $type ) ) {
		return false;
	}

	$meta_group = Tribe__Events__Meta_Factory::get_args( $meta_group_id, $type );
	$meta_ids   = Tribe__Events__Meta_Factory::get_order( $meta_group_id );
	$group_html = '';

	// internal check for hiding items in the meta
	if ( ! $meta_group['show_on_meta'] ) {
		return false;
	}
	$meta_pos_int     = 0;
	$total_meta_items = tribe_count_hierarchical( $meta_ids );
	foreach ( $meta_ids as $meta_id_group ) {
		foreach ( $meta_id_group as $meta_id ) {
			$meta_pos_int ++;

			$group_html = tribe_separated_field( $group_html, $meta_group['wrap']['meta_separator'], tribe_get_meta( $meta_id, $is_the_meta ) );
		}
	}

	$params = [ $meta_group_id ];

	if ( ! empty( $meta['filter_callback'] ) ) {
		return call_user_func_array( $meta['filter_callback'], $params );
	}

	if ( ! empty( $meta['callback'] ) ) {
		$value = call_user_func_array( $meta['callback'], $params );
	}

	$value = empty( $value ) ? $group_html : $value;

	$html = ! empty( $group_html ) ? Tribe__Events__Meta_Factory::template( $meta_group['label'], $value, $meta_group_id, 'meta_group' ) : '';

	return apply_filters( 'tribe_get_meta_group', $html, $meta_group_id );
}

/**
 * Get the meta.
 *
 * @deprecated 4.3
 *
 * @param      $meta_id
 * @param bool $is_the_meta
 *
 * @return bool|mixed|void
 */
function tribe_get_meta( $meta_id, $is_the_meta = false ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	do_action( 'tribe_get_meta', $meta_id, $is_the_meta );

	// die silently if the requested meta item is not registered
	if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id ) ) {
		return false;
	}

	$meta = Tribe__Events__Meta_Factory::get_args( $meta_id );

	// internal check for hiding items in the meta
	if ( ! $meta['show_on_meta'] ) {
		return false;
	}

	$params = [ $meta_id ];

	if ( ! empty( $meta['filter_callback'] ) ) {
		return call_user_func_array( $meta['filter_callback'], $params );
	}

	if ( ! empty( $meta['callback'] ) ) {
		$value = call_user_func_array( $meta['callback'], $params );
	}

	$value = empty( $value ) ? $meta['meta_value'] : $value;

	// if we have a value let's build the html template
	$html = ! empty( $value ) ? Tribe__Events__Meta_Factory::template( $meta['label'], $value, $meta_id ) : '';

	return apply_filters( 'tribe_get_meta', $html, $meta_id );
}

/**
 * Get the args for a meta object.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param        $arg_key
 * @param string $type
 *
 * @return bool
 */
function tribe_get_meta_arg( $meta_id, $arg_key, $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	// die silently if the requested meta group is not registered
	if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id, $type ) ) {
		return false;
	}

	$args = Tribe__Events__Meta_Factory::get_args( $meta_id, $type );

	// check if the arg exists
	if ( isset( $args[ $arg_key ] ) ) {
		return $args[ $arg_key ];
	} else {
		return false;
	}
}

/**
 * Get the template part for the meta object.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param        $template_key
 * @param string $type
 *
 * @return bool
 */
function tribe_get_meta_template_part( $meta_id, $template_key, $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	// die silently if the requested meta group is not registered
	if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id, $type ) ) {
		return false;
	}

	$template = tribe_get_meta_arg( $meta_id, 'wrap', $type );

	if ( isset( $template[ $template_key ] ) ) {
		return $template[ $template_key ];
	} else {
		return false;
	}
}

/**
 * Set the visibility of the meta object.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param bool   $status
 * @param string $type
 */
function tribe_set_the_meta_visibility( $meta_id, $status = true, $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	Tribe__Events__Meta_Factory::set_visibility( $meta_id, $type, $status );
}

/**
 * Set the template for the meta object.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param array  $template
 * @param string $type
 *
 * @return bool
 */
function tribe_set_the_meta_template( $meta_id, $template = [], $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	if ( is_array( $meta_id ) ) {
		foreach ( $meta_id as $id ) {
			tribe_set_the_meta_template( $id, $template, $type );
		}
	} else {
		global $_tribe_meta_factory;

		// die silently if the requested meta group is not registered
		if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id, $type ) ) {
			return false;
		}

		if ( ! empty( $template ) ) {
			$_tribe_meta_factory->{$type}[ $meta_id ]['wrap'] = wp_parse_args( $template, $_tribe_meta_factory->{$type}[ $meta_id ]['wrap'] );
		}
	}

}

/**
 * Set the meta priority to manage positioning.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param int    $priority
 * @param string $type
 *
 * @return bool
 */
function tribe_set_meta_priority( $meta_id, $priority = 100, $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	if ( is_array( $meta_id ) ) {
		foreach ( $meta_id as $id => $priority ) {
			tribe_set_meta_priority( $id, $priority, $type );
		}
	} else {
		global $_tribe_meta_factory;

		// die silently if the requested meta group is not registered
		if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id, $type ) ) {
			return false;
		}

		if ( ! empty( $priority ) ) {
			$_tribe_meta_factory->{$type}[ $meta_id ]['priority'] = $priority;
		}
	}
}

/**
 * Set meta value for meta object.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param        $value
 * @param string $value_type
 * @param string $type
 *
 * @return bool
 */
function tribe_set_meta_value( $meta_id, $value, $value_type = 'meta_value', $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	if ( is_array( $meta_id ) ) {
		foreach ( $meta_id as $id ) {
			tribe_set_meta_value( $id, $value, $value_type, $type );
		}
	} else {
		global $_tribe_meta_factory;

		// die silently if the requested meta group is not registered
		if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id, $type ) ) {
			return false;
		}

		$_tribe_meta_factory->{$type}[ $meta_id ][ $value_type ] = $value;
	}
}

/**
 * Set the meta label for the meta object.
 *
 * @deprecated 4.3
 *
 * @param        $meta_id
 * @param string $label
 * @param string $type
 *
 * @return bool
 */
function tribe_set_meta_label( $meta_id, $label = '', $type = 'meta' ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	if ( is_array( $meta_id ) ) {
		foreach ( $meta_id as $id => $label ) {
			tribe_set_meta_label( $id, $label, $type );
		}
	} else {
		global $_tribe_meta_factory;

		// die silently if the requested meta group is not registered
		if ( ! Tribe__Events__Meta_Factory::check_exists( $meta_id, $type ) ) {
			return false;
		}

		$_tribe_meta_factory->{$type}[ $meta_id ]['label'] = $label;
	}
}

/**
 * Get the event meta
 *
 * @deprecated 4.3
 *
 * @return mixed|void
 */
function tribe_get_the_event_meta() {
	_deprecated_function( __FUNCTION__, '4.3' );

	$html = '';
	foreach ( Tribe__Events__Meta_Factory::get_order() as $meta_groups ) {
		foreach ( $meta_groups as $meta_group_id ) {
			$html .= tribe_get_meta_group( $meta_group_id, true );
		}
	}

	return apply_filters( 'tribe_get_the_event_meta', $html );
}

/**
 * Simple display of meta group tag
 *
 * @deprecated 4.3
 *
 * @uses tribe_get_meta_group()
 * @return echo tribe_get_meta_group( $meta_group_id )
 */
function tribe_display_the_event_meta() {
	_deprecated_function( __FUNCTION__, '4.3' );

	echo apply_filters( 'tribe_display_the_event_meta', tribe_get_the_event_meta() );
}

/**
 *  Simple diplay of meta group tag
 *
 * @uses tribe_get_meta_group()
 * @deprecated 4.3
 *
 * @param string $meta_group_id
 *
 * @return echo tribe_get_meta_group( $meta_group_id )
 */
function tribe_display_meta_group( $meta_group_id ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	echo apply_filters( 'tribe_display_meta_group', tribe_get_meta_group( $meta_group_id ) );
}

/**
 *  Simple diplay of meta tag
 *
 * @uses tribe_get_meta()
 * @deprecated 4.3
 *
 * @param string $meta_id
 *
 * @return echo tribe_get_meta( $meta_id )
 */
function tribe_display_meta( $meta_id ) {
	_deprecated_function( __FUNCTION__, '4.3' );

	echo apply_filters( 'tribe_display_meta', tribe_get_meta( $meta_id ) );
}

/**
 *  Load Legacy Imports
 *
 * @deprecated 4.6.18
 */
if ( ! function_exists( 'Tribe_Events_Importer_Load' ) ) {

	function Tribe_Events_Importer_Load() {
		_deprecated_function( __FUNCTION__, '4.5' );

		Tribe__Events__Importer__Plugin::set_plugin_basename( plugin_basename( __FILE__ ) );
		if ( is_admin() ) {
			add_action( 'init', [ 'Tribe__Events__Importer__Plugin', 'initialize_admin' ], 10, 0 );
			add_action( 'init', [ 'Tribe__Events__Importer__Options', 'process_general_form_submission' ] );
		}
	}

}
