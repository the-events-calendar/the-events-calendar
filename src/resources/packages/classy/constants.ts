export const METADATA_EVENT_URL = '_EventURL';
export const METADATA_EVENT_START_DATE = '_EventStartDate';
export const METADATA_EVENT_END_DATE = '_EventEndDate';
export const METADATA_EVENT_ALLDAY = '_EventAllDay';
export const METADATA_EVENT_TIMEZONE = '_EventTimezone';
export const METADATA_EVENT_ORGANIZER_ID = '_EventOrganizerID';
export const METADATA_EVENT_VENUE_ID = '_EventVenueID';
