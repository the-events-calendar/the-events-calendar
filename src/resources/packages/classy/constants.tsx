// Post types.
export const POST_TYPE_EVENT = 'tribe_events';

// Meta keys.
export const METADATA_EVENT_START_DATE = '_EventStartDate';
export const METADATA_EVENT_END_DATE = '_EventEndDate';
export const METADATA_EVENT_ALLDAY = '_EventAllDay';
export const METADATA_EVENT_TIMEZONE = '_EventTimezone';
