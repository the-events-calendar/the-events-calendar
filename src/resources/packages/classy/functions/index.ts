export * as events from './events';
export * as date from './date';
