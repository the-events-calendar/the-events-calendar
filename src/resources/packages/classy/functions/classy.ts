import {createRoot} from '@wordpress/element';
import {Classy} from "../elements";
import {getElement as getVisualEditorElement, toggleVisibility as toggleVisualEditorVisibility} from './visualEditor';

/**
 * Cached instance of the classy element.
 * Acts as a singleton to avoid creating multiple instances of the same element.
 *
 * @since TBD
 *
 * @type {HTMLElement|null}
 */
let classyElement: HTMLElement | null = null;

/**
 * Returns an existing Classy element or creates a new one if it doesn't exist.
 *
 * This function ensures that only one Classy element exists throughout the application
 * by caching the reference in the `classyElement` variable.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to use for element creation.
 *     If null, the global document will be used.
 * @returns {HTMLElement} The existing or newly created Classy element.
 */
export function getOrCreateElement(document: Document | null = null): HTMLElement {
	document = document ?? window.document;

	if (classyElement !== null) {
		return classyElement;
	}

	const element = document.createElement('div');
	element.id = 'tec-classy';
	element.classList.add('classy-root', 'classy-root--admin');
	classyElement = element;

	return element;
}

/**
 * Inserts the Classy element into the DOM after the visual editor.
 *
 * This function will get or create the Classy element and insert it into the DOM
 * immediately after the visual editor element. If the visual editor cannot be found,
 * the function will return false indicating failure.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to operate within.
 *     If null, the global document will be used.
 *
 * @returns {boolean} True if insertion was successful, false otherwise.
 */
export function insertElement(document: Document | null = null): boolean {
	document = document ?? window.document;
	const element = getOrCreateElement(document);
	const visualEditor = getVisualEditorElement(document);

	if (visualEditor === null) {
		return false;
	}

	visualEditor.parentNode.insertBefore(element, visualEditor.nextSibling);
	toggleVisualEditorVisibility(document);

	return true;
}

/**
 * Toggles the visibility of the Classy element.
 *
 * This function adds or removes the 'classy-root--hidden' class from the Classy element,
 * which should be styled with CSS to show or hide the element accordingly.
 * The element will be created if it doesn't already exist.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to operate within.
 *     If null, the global document will be used.
 *
 * @returns {void}
 */
export function toggleElementVisibility(document: Document | null = null): void {
	document = document ?? window.document;
	getOrCreateElement().classList.toggle('classy-root--hidden');
	toggleVisualEditorVisibility(document);
}

/**
 * Initializes the application by creating and rendering the Classy root element.
 *
 * This function ensures that the Classy element is either retrieved or created
 * and then initializes the React application by rendering the `Classy` component
 * into that element using React's `createRoot` API.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to use for element creation.
 *	 If null, the global document will be used.
 *
 * @returns {void}
 */
export function initApp(document: Document | null = null): void {
	document = document ?? window.document;
	const classyRoot = createRoot(getOrCreateElement(document));
	classyRoot.render(Classy());
}
