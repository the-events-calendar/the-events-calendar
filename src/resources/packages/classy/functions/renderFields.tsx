import React, { Fragment } from 'react';
import { Fill } from '@wordpress/components';
import { PostTitle } from '@tec/common/classy/fields';
import { EventDateTime, EventDetails, EventLocation, EventOrganizer } from '../fields';
import { _x } from '@wordpress/i18n';
import { useSelect } from '@wordpress/data';
import { POST_TYPE_EVENT } from '../constants';

/**
 * Render fields for the Event post type.
 *
 * @since TBD
 *
 * @param {React.ReactNode | null} fields The fields that have been already rendered.
 */
export default function renderFields( fields: React.ReactNode | null ) {
	const postType = useSelect( ( select ) => {
		// @ts-ignore
		return select( 'core/editor' ).getEditedPostAttribute( 'type' );
	}, [] );

	if ( POST_TYPE_EVENT !== postType ) {
		return fields;
	}

	return (
		<Fragment>
			{ /* Render the fields passed to this function first. */ }
			{ fields }

			{/* Portal-render the fields into the Classy form. */ }
			<Fill name="tec.classy.fields">
				<PostTitle
					title={ _x( 'Event Title', 'The title of the event title field.', 'the-events-calendar' ) }
				/>

				<EventDateTime
					title={ _x(
						'Date and Time',
						'The title of the event date and time field.',
						'the-events-calendar'
					) }
				/>

				<EventDetails
					title={ _x( 'Event Details', 'The title of the event details field.', 'the-events-calendar' ) }
				/>

				<EventLocation
					title={ _x( 'Location', 'The title of the event location field.', 'the-events-calendar' ) }
				/>

				<EventOrganizer
					title={ _x( 'Event Organizer', 'The title of the event organizer field.', 'the-events-calendar' ) }
				/>
			</Fill>
		</Fragment>
	);
}
