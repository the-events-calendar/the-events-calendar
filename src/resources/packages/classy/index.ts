import { addAction, addFilter, didAction, doAction } from '@wordpress/hooks';
import renderFields from './functions/renderFields';
import { storeConfig } from './store';
import { STORE_NAME } from './constants';
import { getRegistry } from '@tec/common/classy/store';
import * as constants from './constants';
import { getNewStartEndDates, getAllDayNewDates, getMultiDayDates } from './functions/events';

/**
 * Hook on the Classy application initialization to add TEC store to the Classy registry.
 *
 * @since TBD
 *
 * @return {void} The ECP store is registered.
 */
const registerStore = (): void => {
	getRegistry().registerStore( STORE_NAME, storeConfig );

	/**
	 * Fires after the TEC store is registered and the TEC Classy application is initialized.
	 *
	 * @since TBD
	 *
	 * @return {void} The TEC store is registered.
	 */
	doAction( 'tec.classy.events.initialized' );
};

if ( didAction( 'tec.classy.initialized' ) ) {
	registerStore();
} else {
	addAction( 'tec.classy.initialized', 'tec.classy.events', registerStore );
}

// Hook on the Classy fields rendering logic to render the fields.
addFilter( 'tec.classy.render', 'tec.classy.events', renderFields );

// Export constants and functions to allow their reference in other packages.
export {
	constants,
	getNewStartEndDates,
	getAllDayNewDates,
	getMultiDayDates,
};
