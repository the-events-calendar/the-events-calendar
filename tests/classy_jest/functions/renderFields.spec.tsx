// @ts-nocheck
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference types="jest" />
import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { beforeEach, afterEach, describe, expect, it, jest } from '@jest/globals';
import { Fill, Slot, SlotFillProvider } from '@wordpress/components';
import { POST_TYPE_EVENT } from '@tec/events/classy/constants';
import mockWpDataModule from '../_support/mockWpDataModule';
import TestProvider from '../_support/TestProvider';

// Mock the @wordpress/data module before importing renderFields.
const { mockSelect, mockUseDispatch } = mockWpDataModule();

// Import renderFields after setting up mocks
import { default as renderFields } from '@tec/events/classy/functions/renderFields';

// Mock the field components to avoid complex WordPress component dependencies.
jest.mock( '@tec/common/classy/fields', () => ( {
	PostTitle: ( { title }: { title: string } ) => <div data-testid="post-title">{ title }</div>,
} ) );

// Mock the field components to allow testing their rendering without pulling in complex dependencies.
// The components are tested in their own test files.
jest.mock( '@tec/events/classy/fields', () => ( {
	EventAdmission: ( { title }: { title: string } ) => <div data-testid="event-admission">{ title }</div>,
	EventDateTime: ( { title }: { title: string } ) => <div data-testid="event-datetime">{ title }</div>,
	EventDetails: ( { title }: { title: string } ) => <div data-testid="event-details">{ title }</div>,
	EventLocation: ( { title }: { title: string } ) => <div data-testid="event-location">{ title }</div>,
	EventOrganizer: ( { title }: { title: string } ) => <div data-testid="event-organizer">{ title }</div>,
} ) );

describe( 'renderFields', () => {
	let mockEditPost;

	// Mock currency options.
	const currencyOptions = [
		{ value: 'USD', label: 'US Dollar' },
		{ value: 'EUR', label: 'Euro' },
		{ value: 'GBP', label: 'British Pound' },
	];

	const setupMocks = ( meta = {}, isEvent: boolean = true ) => {
		mockEditPost = jest.fn();

		mockSelect.mockImplementation( ( store: string ): any => {
			if ( store === 'core/editor' ) {
				return {
					getEditedPostAttribute: ( attribute: string ): any => {
						switch ( attribute ) {
							case 'meta':
								return meta;
							case 'type':
								return isEvent ? POST_TYPE_EVENT : 'post';
							case 'title':
								return 'Test Event Title';
							case 'featured_media':
								return 0;
							default:
								return null;
						}
					},
					getEditedPostContent: () => 'Test event content',
					getCurrentPost: () => ( {
						id: 1,
						type: isEvent ? POST_TYPE_EVENT : 'post',
						status: 'publish',
						title: 'Test Event Title',
						content: 'Test event content',
						featured_media: 0,
					} ),
					getCurrentPostType: () => ( isEvent ? POST_TYPE_EVENT : 'post' ),
					getCurrentPostId: () => 1,
				};
			}

			if ( store === 'core/media' ) {
				return {
					getMedia: () => null,
					getMediaItem: () => null,
					getMediaItemByType: () => null,
					getMediaItemsByType: () => [],
				};
			}

			if ( store === 'tec/classy' ) {
				return {
					getCurrencyOptions: () => currencyOptions,
					getDefaultCurrency: () => currencyOptions[ 0 ],
					getVenuesLimit: () => 10,
					getOrganizersLimit: () => 10,
					areTicketsSupported: () => true,
					isUsingTickets: () => false,
				};
			}

			if ( store === 'tec/classy/events' ) {
				return {
					getCurrencyOptions: () => currencyOptions,
					getDefaultCurrency: () => currencyOptions[ 0 ],
					getVenuesLimit: () => 10,
					getOrganizersLimit: () => 10,
					areTicketsSupported: () => true,
					isUsingTickets: () => false,
					isNewEvent: () => false,
					getEventDateTimeDetails: () => ( {
						startDate: '2024-01-01T10:00:00.000Z',
						endDate: '2024-01-01T11:00:00.000Z',
						startTime: '10:00',
						endTime: '11:00',
						allDay: false,
						timezone: 'UTC',
					} ),
					getEventCost: () => ( {
						cost: '',
						currency: 'USD',
						currencyPosition: 'prefix',
						currencySymbol: '$',
					} ),
					getEventOrganizer: () => null,
					getEventVenue: () => null,
				};
			}

			// Throw an error for unknown stores to help identify what needs mocking.
			throw new Error( `Unknown store requested in mockSelect: ${ store }` );
		} );

		mockUseDispatch.mockImplementation( ( store: string ): any => {
			if ( store === 'core/editor' ) {
				return {
					editPost: mockEditPost,
				};
			}

			if ( store === 'core/media' ) {
				return {
					uploadMedia: jest.fn(),
					createMediaFromFile: jest.fn(),
					editMedia: jest.fn(),
					deleteMedia: jest.fn(),
				};
			}

			if ( store === 'tec/classy/events' ) {
				return {
					updateEventMeta: jest.fn(),
					setEventDateTimeDetails: jest.fn(),
					setEventCost: jest.fn(),
					setEventOrganizer: jest.fn(),
					setEventVenue: jest.fn(),
				};
			}

			// Throw an error for unknown stores to help identify what needs mocking.
			throw new Error( `Unknown store requested in mockUseDispatch: ${ store }` );
		} );
	};

	beforeEach( () => {
		jest.resetModules();
		jest.clearAllMocks();
	} );

	afterEach( () => {
		jest.resetModules();
	} );

	it( 'renders nothing if post is not Event', () => {
		setupMocks( {}, false );

		const TestComponent = () => renderFields( null );

		const { container } = render(
			<TestProvider>
				<SlotFillProvider>
					<TestComponent />
					<Slot name="tec.classy.fields" />
				</SlotFillProvider>
			</TestProvider>
		);

		expect( container ).toMatchSnapshot();

		// Check that the main elements are not rendered.
		expect( () => screen.getByTestId( 'post-title' ) ).toThrowError();
		expect( () => screen.getByTestId( 'event-datetime' ) ).toThrowError();
		expect( () => screen.getByTestId( 'event-details' ) ).toThrowError();
		expect( () => screen.getByTestId( 'event-location' ) ).toThrowError();
		expect( () => screen.getByTestId( 'event-organizer' ) ).toThrowError();
		expect( () => screen.getByTestId( 'event-admission' ) ).toThrowError();
	} );

	it( 'renders all fields if post is Event', () => {
		setupMocks();

		const TestComponent = () => renderFields( null );

		const { container } = render(
			<TestProvider>
				<SlotFillProvider>
					<TestComponent />
					<Slot name="tec.classy.fields" />
				</SlotFillProvider>
			</TestProvider>
		);

		expect( container ).toMatchSnapshot();

		// Check that the main elements are rendered. More detailed tests for each field are in their own test files.
		expect( screen.getByTestId( 'post-title' ) ).toBeInTheDocument();
		expect( screen.getByTestId( 'event-datetime' ) ).toBeInTheDocument();
		expect( screen.getByTestId( 'event-details' ) ).toBeInTheDocument();
		expect( screen.getByTestId( 'event-location' ) ).toBeInTheDocument();
		expect( screen.getByTestId( 'event-organizer' ) ).toBeInTheDocument();
		expect( screen.getByTestId( 'event-admission' ) ).toBeInTheDocument();
	} );

	it( 'renders additional fields added via the tec.classy.fields Slot', () => {
		setupMocks();

		const TestComponent = () => renderFields( null );

		const { container } = render(
			<TestProvider>
				<SlotFillProvider>
					<TestComponent />
					<Fill name="tec.classy.fields">
						<p>Additional Field 1</p>
						<p>Additional Field 2</p>
					</Fill>
					<Slot name="tec.classy.fields" />
				</SlotFillProvider>
			</TestProvider>
		);

		expect( container ).toMatchSnapshot();

		// Check that the additional fields are rendered.
		expect( screen.getByText( 'Additional Field 1' ) ).toBeInTheDocument();
		expect( screen.getByText( 'Additional Field 2' ) ).toBeInTheDocument();
	} );
} );
