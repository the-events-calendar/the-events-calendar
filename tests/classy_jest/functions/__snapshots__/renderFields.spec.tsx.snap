// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderFields renders additional fields added via the tec.classy.fields Slot 1`] = `
<div>
  <div
    data-testid="post-title"
  >
    Event Title
  </div>
  <div
    data-testid="event-datetime"
  >
    Date and Time
  </div>
  <div
    data-testid="event-details"
  >
    Event Details
  </div>
  <div
    data-testid="event-location"
  >
    Location
  </div>
  <div
    data-testid="event-organizer"
  >
    Event Organizer
  </div>
  <div
    data-testid="event-admission"
  >
    Admission
  </div>
  <p>
    Additional Field 1
  </p>
  <p>
    Additional Field 2
  </p>
</div>
`;

exports[`renderFields renders all fields if post is Event 1`] = `
<div>
  <div
    data-testid="post-title"
  >
    Event Title
  </div>
  <div
    data-testid="event-datetime"
  >
    Date and Time
  </div>
  <div
    data-testid="event-details"
  >
    Event Details
  </div>
  <div
    data-testid="event-location"
  >
    Location
  </div>
  <div
    data-testid="event-organizer"
  >
    Event Organizer
  </div>
  <div
    data-testid="event-admission"
  >
    Admission
  </div>
</div>
`;

exports[`renderFields renders nothing if post is not Event 1`] = `<div />`;
